oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/authentication/IntrospectedOAuth2TokenAuthenticationToken.java

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright 2002-2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.security.oauth2.server.resource.authentication;
+
+import java.util.Collection;
+import java.util.Map;
+
+import org.springframework.security.core.GrantedAuthority;
+import org.springframework.security.core.SpringSecurityCoreVersion;
+import org.springframework.security.oauth2.core.OAuth2TokenWithClaims;
+
+/**
+ * An {@link org.springframework.security.core.Authentication} token that represents a successful authentication as
+ * obtained through an opaque token
+ * <a target="_blank" href="https://tools.ietf.org/html/rfc7662">introspection</a>
+ * process.
+ *
+ * @author Josh Cummings
+ * @since 5.2
+ */
+public class IntrospectedOAuth2TokenAuthenticationToken extends AbstractOAuth2TokenAuthenticationToken<OAuth2TokenWithClaims> {
+	private static final long serialVersionUID = SpringSecurityCoreVersion.SERIAL_VERSION_UID;
+
+	public IntrospectedOAuth2TokenAuthenticationToken(OAuth2TokenWithClaims token) {
+		super(token);
+	}
+
+	public IntrospectedOAuth2TokenAuthenticationToken(OAuth2TokenWithClaims token,
+			Collection<? extends GrantedAuthority> authorities) {
+
+		super(token, authorities);
+		setAuthenticated(true);
+	}
+
+	@Override
+	public Map<String, Object> getTokenAttributes() {
+		return getToken().getClaims();
+	}
+
+	@Override
+	public String getName() {
+		return (String) getTokenAttributes().get("username");
```

Comment:
Again it is strange this is overridden vs using the principal toString

Reply:
Good point, I'll think about this a bit more. I'll just mention here that this is similar to [`JwtAuthenticationToken`](https://github.com/spring-projects/spring-security/blob/master/oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/authentication/JwtAuthenticationToken.java#L72), which was the main driver for doing it this way.

Reply:
It would be odd, I think, to call `getPrincipal().toString()` here because `getPrincipal()` is an `AbstractOAuth2Token`.

It would mean that either `AbstractOAuth2Token` (or its derivative for this class) would need to pick a claim from itself to be the principal's "name", and that feels strange since an OAuth token is not inherently a "named" principal.

Perhaps it would be better to add `String name` as a constructor parameter to this authentication token.

Reply:
Something else to consider is that `username` is an optional attribute. So is `sub` but that would be a more logical choice I think. Additionally, depending on the provider, the subject identifier may be a non-standard attribute. This may need to be customized.

Reply:
From the spec, `sub` is:

> Usually a machine-readable identifier of the resource owner who authorized this token.

And `username` is:

> Human-readable identifier for the resource owner who authorized this token.

There isn't any indication in `javax.security.Principal` as to whether it is intended to be human or machine-readable, but it seems to me that if `getPrincipal().toString()` is usually the clearer choice, then I'd expect it to be the human-readable one.

What makes you say that `sub` is the more logical choice?

Reply:
> What makes you say that sub is the more logical choice?

`sub` should be a GUID when combined with `iss`. However, it's possible that `username` may not be unique since 2 people can have the same name/username across systems (`iss`).

Reply:
For consistency with other flows, let's keep with `sub`.